src/IO/Importers/VolumeDataReaderActions.hpp

```diff
@@ -83,115 +88,130 @@ namespace ThreadedActions {
+      typename ArrayIndex,
+      Requires<db::tag_is_retrievable_v<Tags::RegisteredElements, DataBox> and
+               db::tag_is_retrievable_v<Tags::HasReadVolumeData, DataBox>> =
+          nullptr>
+  static void apply(DataBox& box, Parallel::GlobalCache<Metavariables>& cache,
+                    const ArrayIndex& /*array_index*/) noexcept {
+    // Only read and distribute the volume data once
+    // This action will be invoked by `importers::Actions::ReadVolumeData` from
+    // every element on the node, but only the first invocation reads the file
+    // and distributes the data to all elements. Subsequent invocations do
+    // nothing. We use the `ImporterOptionsGroup` that specifies the data file
+    // to read in as the identifier for whether or not we have already read the
+    // requested data. Doing this at runtime avoids having to collect all
+    // data files that will be read in at compile-time to initialize a flag in
+    // the DataBox for each of them.
+    const auto& has_read_volume_data = db::get<Tags::HasReadVolumeData>(box);
+    const auto volume_data_id = pretty_type::get_name<ImporterOptionsGroup>();
+    if (has_read_volume_data.find(volume_data_id) !=
+        has_read_volume_data.end()) {
+      return;
+    }
+    db::mutate<Tags::HasReadVolumeData>(
+        make_not_null(&box),
+        [&volume_data_id](const gsl::not_null<std::unordered_set<std::string>*>
+                              local_has_read_volume_data) noexcept {
+          local_has_read_volume_data->insert(std::move(volume_data_id));
+        });
 
-  template <typename ParallelComponent, typename DataBox,
-            typename Metavariables, typename ArrayIndex,
-            Requires<db::tag_is_retrievable_v<Tags::RegisteredElements,
-                                              DataBox>> = nullptr>
-  static void apply(
-      DataBox& box, Parallel::GlobalCache<Metavariables>& cache,
-      const ArrayIndex& /*array_index*/,
-      const gsl::not_null<Parallel::NodeLock*> node_lock) noexcept {
-    node_lock->lock();
-    {
-      // The scoping is to close the file before unlocking
-      h5::H5File<h5::AccessType::ReadOnly> h5file(
-          Parallel::get<Tags::FileName<ImporterOptionsGroup>>(cache));
-      constexpr size_t version_number = 0;
-      const auto& volume_file = h5file.get<h5::VolumeData>(
-          "/" + Parallel::get<Tags::Subgroup<ImporterOptionsGroup>>(cache),
-          version_number);
-      const auto observation_id = volume_file.find_observation_id(
-          Parallel::get<Tags::ObservationValue<ImporterOptionsGroup>>(cache));
-      // Read the tensor data for all elements at once, since that's how it's
-      // stored in the file
-      tuples::tagged_tuple_from_typelist<FieldTagsList> all_tensor_data{};
-      tmpl::for_each<FieldTagsList>([
-        &all_tensor_data, &volume_file, &observation_id
-      ](auto field_tag_v) noexcept {
+    // Open the volume data file
```

Comment:
Code between here and line 206 is unchanged. It only shows up in this diff because the indentation level is reduced by two spaces.